lib/setopt.c

```diff
@@ -77,6 +77,51 @@ CURLcode Curl_setstropt(char **charp, const char *s)
   return CURLE_OK;
 }
 
+CURLcode Curl_setblobopt(struct curl_blob **blobp,
+                         const struct curl_blob *st_blobs)
+{
+  /* Release the previous storage at `blobp' and replace by a dynamic storage
+     copy of `st_blobs'.
+     if flags matches CURL_BLOB_DUPFLAG_COPY, the data are copied after the
+     memory blob.
+     if flags not matches CURL_BLOB_DUPFLAG_COPY, only the pointer is copied.
+     memory blob.
+     Return CURLE_OK, CURLE_BAD_FUNCTION_ARGUMENT or CURLE_OUT_OF_MEMORY. */
+
+  Curl_safefree(*blobp);
+
+  if(st_blobs) {
+    /* sblob contain struct curl_blob data */
+
+    /* dup_pos contain offset of copied data in allocated buffer
+       dup_pos is aligned on a 0x10 boundary, because memcpy
+       (and pehaps sometime buffer usage) is faster with aligned data
+       on somes architectures */
+    /* Note : data between sizeof(struct curl_blob) and dup_pos
+       are not used and not initialised */
+    size_t dup_pos = ((st_blobs->flags) & CURL_BLOB_DUPFLAG_COPYMASK) ?
+      (((sizeof(struct curl_blob))>>4)<<4) + 0x10 : 0;
+    struct curl_blob *st_blob_dup = (struct curl_blob *)
+        malloc(sizeof(struct curl_blob) +
+               dup_pos + (dup_pos ? st_blobs->len : 0));
+    if(!st_blob_dup)
+        return CURLE_OUT_OF_MEMORY;
+    *st_blob_dup = *st_blobs;
+    if(dup_pos) {
+      /* dupflag is set by on blob. We need copy the data,
+      *   app don't need keep data pointer valid after setopt */
+      st_blob_dup->data = (void *)(((char *)st_blob_dup) +
+                                   dup_pos);
+      memcpy(st_blob_dup->data, st_blobs->data, st_blobs->len);
+    }
```

Comment:
Can I propose we simplify this function? It is written to look more scary than it is in reality. My proposal:
~~~c
{
  struct curl_blob *nblob;
  /* allocate memory for blob + possibly the data */
  size_t memsize = sizeof(struct curl_blob) +
    (st_blobs->flags & CURL_BLOG_COPY ? st_blobs->len : 0);
  nblob = malloc(memsize);
  if(!nlob)
    return CURLE_OUT_OF_MEMORY;
  *nblob = *st_blobs;
  if(st_blobs->flags & CURL_BLOG_COPY) {
    /* put the data after the blob struct in memory */
    char *nptr = (char *)nblob + sizeof(struct curl_blob);
    memcpy(nptr, st_blobs->data, st_blobs->len);
    nblob->data = nptr;
  }
}
~~~

Another thing that struck me. I think this function should also return an error if the set blob seems to be too large. We do this for all data options avoid integer overflow risks and to just stop mistakes or abuse earely. Something like this:
~~~c
  if(st_blobs->len > CURL_MAX_INPUT_LENGTH)
    return CURLE_BAD_FUNCTION_ARGUMENT;

Reply:
I rewrote the function in the spirit of your code, but I still align new data.
I really think memcpy is sometime faster with aligned data, and it is better align somes binary structure. Of course, this change nothing on PEM text..

I known there is two discussion between us (copy option, and alignement)
Of course, at end I'll accept your decision :-)

Reply:
Such memcpy implementations first copy the few bytes first that arn't aligned, then copy the rest aligned as fast as it can (see for example glibc's implementation). Our job is to write libcurl code as simple and easy to read as possible. Not second-guessing function calls. For these reasons, I don't think insisting on alignment here is a good idea.

Reply:
ok, but I'll use int a flags type, so at least adress will not be odd.
there is a lot of "int flags" in curl code.

int is 4 bytes on all platform I test, so we will not have ABI compatibility prolem

I pushed an additional commit which remove alignement code

Reply:
About alignement, somes link
https://lemire.me/blog/2012/05/31/data-alignment-for-speed-myth-or-reality/

https://blog.quarkslab.com/unaligned-accesses-in-cc-what-why-and-solutions-to-do-it-properly.html



I don't have personal experience on alignement. I had only crash on old misc processor when I read a 4 or 8 bytes integer on odd address.

another alternative can be
#define ROUND_INTEGER_0x10(x)  ((((x)+0x0f) / 0x10) * 0x10)

and then just replace the two occurence of
sizeof(struct curl_blob)
by
ROUND_INTEGER_0x10(sizeof(struct curl_blob))

for me the is not excessive cost of complexity

Reply:
**No manual alignment at all**. Get rid of it.

If you can show *any* speed difference at all between an aligned approach an unaligned when you do this perhaps one copy per transfer, then please go ahead. On any architecture of your choice. Doing *any* kind of manual optimization like this before you can show even a tiny teeny performance benefit is just *wrong* and we call it *premature optimization*.

Build this code on a typical x86-64 with gcc:
~~~c
struct test {
  char *hello;
  char byte;
};

void main()
{
  printf("size: %d\n", sizeof(struct test));
}
~~~
Then run it:
~~~
$ ./a.out 
size: 16

Reply:
same thing with Visual studio 2019, (except if we add before #pragma pack(1)

ok, as gcc and VS2019 does themself the around 8 work, I accept your simplification :-)